sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala

```
@@ -444,6 +447,156 @@ case class UnresolvedStar(target: Option[Seq[String]]) extends Star with Unevalu
+
+      // map input columns while searching for the except entry corresponding to the current column
+      columns.map(col => col -> groupedExcepts.get(col.toAttribute)).collect {
+        // pass through columns that don't match anything in groupedExcepts
+        case (col, None) => col
+        // found a match but nestedExcepts has remaining excepts - recurse to rewrite the struct
+        case (col, Some(nestedExcepts)) if !nestedExcepts.exists(_.isEmpty) =>
+          val fields = col.dataType match {
+            case s: StructType => s.fields
+            // we shouldn't be here since we throw the same error above (in getRootStruct), but
+            // nonetheless just throw the same error
+            case _ => throw new AnalysisException(
+              errorClass = "EXCEPT_NESTED_COLUMN_INVALID_TYPE",
+              messageParameters =
+                Map("columnName" -> col.qualifiedName, "dataType" -> col.dataType.toString))
+          }
+          val extractedFields = fields.zipWithIndex.map { case (f, i) =>
+            Alias(GetStructField(col, i), f.name)()
+          }
+          val newExcepts = nestedExcepts.map { nestedExcept =>
+            // INVARIANT: we cannot have duplicate column names in nested columns, thus, this `head`
+            // will find the one and only column corresponding to the correct extractedField.
+            extractedFields.collectFirst { case col if resolver(col.name, nestedExcept.head) =>
+              col.toAttribute -> nestedExcept.tail
+            }.get
+          }
+          Alias(CreateStruct(filterColumns(extractedFields.toSeq, newExcepts)), col.name)()
+        // if there are multiple nestedExcepts but one is empty we must have overlapping except
+        // columns. throw an error.
+        case (col, Some(nestedExcepts)) if nestedExcepts.size > 1 =>
+          throw new AnalysisException(errorClass = "EXCEPT_OVERLAPPING_COLUMNS", Map.empty)
```

it's not overlapping but duplicated?

---

overlapping is more accurate as this error is thrown even with overlapping columns in nested structs - e.g.
select * except(data.a.b, data.a) should throw this exception as well

---

oh I see!